onmt/models/model.py

```diff
@@ -100,6 +101,64 @@ def load_state_dict(
                 % checkpoint["generator"].keys()
             )
 
+    def load_safe_state_dict(
+        self,
+        model_path,
+        precision=torch.float32,
+        device=torch.device("cpu"),
+        strict=True,
+    ):
+        """Custom state_dict loading to enable moving module on device as they are loaded
+
+        Args:
+            checkpoint:
+            precision:
+            device:
+            strict:
+
+        Return:
+
+            * Model"""
+        # bitsandbytes quantize weights when .cuda() is called
+        # for huge models we need to save Ram
+        # so we load the weights  module by module and transfer them to GPU for quantization
+        try:
+            import safetensors
+        except ImportError:
+            raise ImportError("run: pip install safetensors, to use safetensors")
+        keyfound = {}
+        shards = glob.glob(model_path + ".*.safetensors")
+        f = []
+        keys_shard = {}
+        for i, shard in enumerate(shards):
+            f.append(safetensors.safe_open(shard, framework="pt", device="cpu"))
```

Comment:
device -> use arg passed in the function?

Reply:
no, the way it works is:
each module (the whole model) is built empty on cpu. we load the data of each module on cpu, and as each module is loaded with the tensors, it is moved to gpu (this is exactly at this step that data is quantized).
see the module.to(device) juste after.